File: solarsystem/src/main/java/fail/stderr/usb/common/FrameFunctions.java
package fail.stderr.usb.common;

import org.orekit.frames.Frame;
import org.orekit.orbits.KeplerianOrbit;

import java.lang.reflect.InvocationTargetException;

public class FrameFunctions {

  public static Frame createRootFrame(String name) {
    try {
      var c = Frame.class.getDeclaredConstructor(String.class, boolean.class);
      c.setAccessible(true);
      var frame = c.newInstance(name, true);
      return frame;
    } catch (NoSuchMethodException | InstantiationException | IllegalAccessException | InvocationTargetException e) {
      throw new RuntimeException(e);
    }
  }

  public static Frame createBodyFrame(String bodyName, KeplerianOrbit bodyOrbit, Frame parentFrame) {
    var transformProvider = new BodyToParentTransformProvider(bodyOrbit, parentFrame);
    return new Frame(parentFrame, transformProvider, "%sCenteredFrame".formatted(bodyName), true);
  }

}

File: solarsystem/src/main/kotlin/fail/stderr/usb/common/FrameFunctions.kt
package fail.stderr.usb.common

import org.orekit.frames.Frame
import org.orekit.orbits.KeplerianOrbit

class FrameFunctions {

  companion object {

    @JvmStatic
    fun createRootFrame(name: String?): Frame {
      val c = Frame::class.java.getDeclaredConstructor(
        String::class.java, Boolean::class.javaPrimitiveType
      )
      c.isAccessible = true
      val frame = c.newInstance(name, true)
      return frame
    }

    @JvmStatic
    fun createBodyFrame(bodyName: String?, bodyOrbit: KeplerianOrbit?, parentFrame: Frame?): Frame {
      val transformProvider = BodyToParentTransformProvider(bodyOrbit!!, parentFrame!!)
      return Frame(parentFrame, transformProvider, "${bodyName}CenteredFrame", true)
    }

  }

}
